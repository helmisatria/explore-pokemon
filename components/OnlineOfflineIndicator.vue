<template>
  <transition name="slide-fade">
    <div
      v-show="isShowIndicator || onlineStatus === 'offline'"
      class="online-status-indicator max-w-lg sticky top-0 z-10 transition-all text-white text-center text-sm bg-opacity-90"
      :class="onlineStatus === 'online' ? 'bg-green-600' : 'bg-gray-800'"
    >
      <p v-if="onlineStatus === 'offline'" class="py-2">
        You're offline, please check your connection
      </p>
      <p v-else-if="onlineStatus === 'online'" class="py-2">You're online, welcome back!</p>
    </div>
  </transition>
</template>

<script>
import { mapState } from 'pinia'
import { usePokemonStore } from '~/store'

export default {
  name: 'OnlineOfflineIndicator',
  data() {
    return {
      isShowIndicator: false,
    }
  },
  computed: {
    ...mapState(usePokemonStore, ['onlineStatus']),
  },
  mounted() {
    this.init()
  },
  methods: {
    init() {
      const store = usePokemonStore()

      if (!process.client) return

      if (!navigator.onLine) {
        this.isShowIndicator = true
        store.onlineStatus = 'offline'
      }

      window.addEventListener('online', () => {
        this.isShowIndicator = true
        store.onlineStatus = 'online'

        setTimeout(() => {
          this.isShowIndicator = false
          store.onlineStatus = 'idle'
        }, 3000)
      })

      window.addEventListener('offline', () => {
        this.isShowIndicator = true
        store.onlineStatus = 'offline'
      })
    },
  },
}
</script>

<style scoped>
.slide-fade-enter-active {
  transition: all 0.5s ease;
}
.slide-fade-leave-active {
  transition: all 0.5s ease;
}
.slide-fade-enter, .slide-fade-leave-to
/* .slide-fade-leave-active below version 2.1.8 */ {
  transform: translateY(-10px);
  opacity: 0;
}
</style>
