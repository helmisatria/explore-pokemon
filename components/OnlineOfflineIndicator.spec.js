import { createLocalVue, shallowMount } from '@vue/test-utils'
import { PiniaVuePlugin } from 'pinia'
import { createTestingPinia } from '@pinia/testing'
import OnlineOfflineIndicator from './OnlineOfflineIndicator.vue'

describe('Component: OnlineOfflineIndicator', () => {
  let wrapper

  process.client = true

  const localVue = createLocalVue()
  localVue.use(PiniaVuePlugin)

  delete window.navigator
  window.navigator = { onLine: false }

  const generateWrapper = () => {
    return shallowMount(OnlineOfflineIndicator, {
      localVue,
      pinia: createTestingPinia(),
      mocks: {
        $router: { go: jest.fn() },
      },
    })
  }

  beforeEach(() => {
    wrapper = generateWrapper()
  })

  it('should mount component properly', () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  describe('Handle Indicator Visibility', () => {
    it('should display indicator, given status is offline', async () => {
      wrapper.vm.$pinia.state.value.pokemon.onlineStatus = 'offline'
      await wrapper.vm.$nextTick()

      expect(wrapper.find('.online-status-indicator').element.style.display).toBe('')
    })

    it('should display indicator, given status is not idle', async () => {
      wrapper.vm.$pinia.state.value.pokemon.onlineStatus = 'online'
      await wrapper.vm.$nextTick()

      expect(wrapper.find('.online-status-indicator').element.style.display).toBe('')
    })

    it('should NOT display indicator, given status is idle', async () => {
      window.navigator = { onLine: true }
      wrapper = generateWrapper()

      await wrapper.setData({ isShowIndicator: false })
      wrapper.vm.$pinia.state.value.pokemon.onlineStatus = 'idle'
      await wrapper.vm.$nextTick()

      expect(wrapper.find('.online-status-indicator').element.style.display).toBe('none')
    })
  })

  describe('Handle Copy Inside Indicator', () => {
    it('should display text inside indicator properly based on the status', async () => {
      wrapper = generateWrapper()

      wrapper.vm.$pinia.state.value.pokemon.onlineStatus = 'offline'
      await wrapper.vm.$nextTick()

      expect(wrapper.find('.online-status-indicator').text()).toBe(
        `You're offline, please check your connection`
      )

      wrapper.setData({ isShowIndicator: true })
      wrapper.vm.$pinia.state.value.pokemon.onlineStatus = 'online'
      await wrapper.vm.$nextTick()

      expect(wrapper.find('.online-status-indicator').text()).toBe(`You're online, welcome back!`)
    })
  })
})
