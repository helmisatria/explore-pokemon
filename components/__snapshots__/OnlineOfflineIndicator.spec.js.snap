// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: OnlineOfflineIndicator should mount component properly 1`] = `
<transition-stub
  name="slide-fade"
>
  <div
    class="online-status-indicator max-w-lg sticky top-0 z-10 transition-all text-white text-center text-sm bg-opacity-90 bg-gray-800"
    style=""
  >
    <p
      class="py-2"
    >
      
      You're offline, please check your connection
    
    </p>
  </div>
</transition-stub>
`;
